teps/0052-tekton-results-automated-run-resource-cleanup.md

```diff
@@ -0,0 +1,213 @@
+---
+status: proposed
+title: "Tekton Results: Automated Run Resource Cleanup"
+creation-date: "2021-02-11"
+last-updated: "2021-02-11"
+authors:
+  - "@wlynch"
+---
+
+# TEP-0052: Tekton Results: Automated Run Resource Cleanup
+
+<!-- toc -->
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Delete Runs when complete](#delete-runs-when-complete)
+  - [Grace Period](#grace-period)
+  - [Notes/Caveats (optional)](#notescaveats-optional)
+    - [Future Work](#future-work)
+      - [Grace Period API Configuration](#grace-period-api-configuration)
+  - [Risks and Mitigations](#risks-and-mitigations)
+    - [Early Deletion](#early-deletion)
+    - [Watcher Outages](#watcher-outages)
+  - [User Experience (optional)](#user-experience-optional)
+  - [Performance (optional)](#performance-optional)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Alternatives](#alternatives)
+  - [Decouple Cleanup from Watcher](#decouple-cleanup-from-watcher)
+
+<!-- /toc -->
+
+## Summary
+
+This proposal details a solution to provide automatic TaskRun/PipelineRun
+resource clean up in conjunction with Tekton Results to preserve long term
+history while minimizing the footprint of data the Pipeline controller needs to
+keep track of.
+
+## Motivation
+
+Today, most users view Tekton Run history via the Tekton Pipeline API (e.g.
+kubectl, tkn, dashboard, etc). However since all CRD objects are
+[stored in etcd](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/),
+this creates a problem for the Pipeline controller - as the build history grows
+over time, so does the corpus of data the controller needs to handle in its data
+plane. Over time this grows unbounded, significantly degrading the performance
+of the cluster as more Runs are completed.
```

Comment:
maybe a silly question, but why is the controller continuing to be informed about objects which have been completed?

(i think if the controller was restarted it would try to re-process _all_ of them which would be terrible, but wondering if this problem only impacts the controller on restart or at other times too)

Reply:
nice!! thanks for doing the additional investigation